kedro/runner/runner.py

```diff
@@ -157,31 +158,121 @@ def create_default_data_set(self, ds_name: str) -> AbstractDataSet:
-            "You can resume the pipeline run by adding the following "
-            "argument to your previous command:\n%s",
-            len(remaining_nodes),
-            postfix,
-        )
+        if not postfix:
+            self._logger.warning(
+                "No nodes ran. Repeat the previous command to attempt a new run."
+            )
+        else:
+            self._logger.warning(
+                "There are %d nodes that have not run.\n"
+                "You can resume the pipeline run from the nearest nodes with "
+                "persisted inputs by adding the following "
+                "argument to your previous command:\n%s",
+                len(remaining_nodes),
+                postfix,
+            )
+
+
+def _find_persistent_ancestors(
+    pipeline: Pipeline, children: Iterable[Node], catalog: DataCatalog
+) -> Set[Node]:
+    """Breadth-first search approach to finding the complete set of
+    persistent ancestors of an iterable of ``Node``s. Persistent
+    ancestors exclusively have persisted ``Dataset``s as inputs.
+
+    Args:
+        pipeline: the ``Pipeline`` to find ancestors in.
+        children: the iterable containing ``Node``s to find ancestors of.
+        catalog: the ``DataCatalog`` of the run.
+
+    Returns:
+        A set containing first persistent ancestors of the given
+        ``Node``s.
+
+    """
+    ancestor_nodes_to_run = set()
+    queue, visited = deque(children), set()
+    while queue:
+        current_node = queue.popleft()
+        visited.add(current_node)
+        if _has_persistent_inputs(current_node, catalog):
+            ancestor_nodes_to_run.add(current_node)
+            continue
+        for parent in _enumerate_parents(pipeline, current_node):
+            if parent not in visited:
+                queue.append(parent)
+    return ancestor_nodes_to_run
+
+
+def _enumerate_parents(pipeline: Pipeline, child: Node) -> List[Node]:
+    """For a given ``Node``, returns a list containing the direct parents
+    of that ``Node`` in the given ``Pipeline``.
+
+    Args:
+        pipeline: the ``Pipeline`` to search for direct parents in.
+        child: the ``Node`` to find parents of.
+
+    Returns:
+        A list of all ``Node``s that are direct parents of ``child``.
+
+    """
+    parent_pipeline = pipeline.only_nodes_with_outputs(*child.inputs)
+    return parent_pipeline.nodes
+
+
+def _has_persistent_inputs(node: Node, catalog: DataCatalog) -> bool:
+    """Check if a ``Node`` exclusively has persisted Datasets as inputs.
+    If at least one input is a ``MemoryDataSet``, return False.
+
+    Args:
+        node: the ``Node`` to check the inputs of.
+        catalog: the ``DataCatalog`` of the run.
+
+    Returns:
+        True if the ``Node`` being checked exclusively has inputs that
+        are not ``MemoryDataSet``, else False.
+
+    """
+    for node_input in node.inputs:
+        if isinstance(vars(catalog.datasets)[node_input], MemoryDataSet):
```

Comment:
Why do you need to do `vars(catalog.datasets)`?

Reply:
catalog.datasets is an object of FrozenDatasets, which has no __getitem__ defined for it. It stores datasets in the __dict__ attribute, which can be accessed in a readable way by using vars().

The alternative would be catalog.datasets.__dict__, or to use the protected catalog._data_sets (which is already a dict) but linters won't allow these.

Reply:
TIL! I've literally never used `vars` before so this is really interesting to hear! 👏 

That said, the way I would recommend doing this would actually be `._get_dataset(dataset_name)` and ignore any lint error it might give. This is a bit more obvious and would be consistent with[ how it's done in kedro-viz](https://github.com/kedro-org/kedro-viz/blob/1d7f37eab55fb12f596f5581997b9fbffefd8735/package/kedro_viz/data_access/repositories/catalog.py#L55).

Reply:
P.S. yes, it is kind of ridiculous we don't have a nicer way of doing this. It's one thing I would love to improve: see the first bullet point in https://github.com/kedro-org/kedro/issues/1778.

Reply:
@AntonyMilneQB but isn't the problem here that you don't have the actual dataset names?

Reply:
oh! yeah probably! I keep getting so confused by this part of the code base 😅

Reply:
> That said, the way I would recommend doing this would actually be `catalog._get_dataset(dataset_name)` and ignore any lint error it might give.

Thanks for the feedback! Yeah ``vars()`` is quite uncommon I feel 😁. I made this change in [7c7e795](https://github.com/kedro-org/kedro/pull/1795/commits/7c7e79546c2739d034d03444a00dc0fdb6304414).